diff --git a/arch/x86/include/asm/unistd_64.h b/arch/x86/include/asm/unistd_64.h
index 0431f19..a44956e 100644
--- a/arch/x86/include/asm/unistd_64.h
+++ b/arch/x86/include/asm/unistd_64.h
@@ -686,6 +686,10 @@ __SYSCALL(__NR_getcpu, sys_getcpu)
 __SYSCALL(__NR_process_vm_readv, sys_process_vm_readv)
 #define __NR_process_vm_writev			311
 __SYSCALL(__NR_process_vm_writev, sys_process_vm_writev)
+#define __NR_multikernel_boot                   312
+__SYSCALL(__NR_multikernel_boot, sys_multikernel_boot)
+#define __NR_get_boot_params_addr               313
+__SYSCALL(__NR_get_boot_params_addr, sys_get_boot_params_addr)
 
 #ifndef __NO_STUBS
 #define __ARCH_WANT_OLD_READDIR