diff --git a/drivers/gpu/vga/vgaarb.c b/drivers/gpu/vga/vgaarb.c
index 111d956..03822da 100644
--- a/drivers/gpu/vga/vgaarb.c
+++ b/drivers/gpu/vga/vgaarb.c
@@ -1304,6 +1304,11 @@ static int __init vga_arb_device_init(void)
 	struct pci_dev *pdev;
 	struct vga_device *vgadev;
 
+	/* POPCORN -- not sure
+	 * TODO -- check this! */
+	if ( !is_bsp_cluster )
+		return -1;
+
 	rc = misc_register(&vga_arb_device);
 	if (rc < 0)
 		pr_err("vgaarb: error %d registering device\n", rc);