diff --git a/drivers/acpi/processor_driver.c b/drivers/acpi/processor_driver.c
index 9d7bc9f..89a640b 100644
--- a/drivers/acpi/processor_driver.c
+++ b/drivers/acpi/processor_driver.c
@@ -469,7 +469,13 @@ static int __cpuinit acpi_processor_add(struct acpi_device *device)
 	}
 
 #ifdef CONFIG_SMP
-	if (pr->id >= setup_max_cpus && pr->id != 0)
+	printk(KERN_INFO
+			"pr->id %d setup_max_cpus %d nr_cpu_ids %d cpu_present %d boot %d prev_check %d curr_check %d\n",
+			pr->id, setup_max_cpus, nr_cpu_ids, cpu_present(pr->id), boot_cpu_physical_apicid,
+			(pr->id >= setup_max_cpus && pr->id != 0), ( !cpu_present(pr->id) && pr->id != first_cpu(cpu_present_map)));
+//	if (pr->id >= setup_max_cpus && pr->id != 0)
+	if (!cpu_present(pr->id) 
+           && (pr->id != first_cpu(cpu_present_map)))
 		return 0;
 #endif
 